Richify the CLI

alex-zenml opened the pull request:
## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [x] If my change requires a change to the documentation, I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Describe changes
- In general, adds `rich` features to our CLI output
- All tracebacks are now `rich` tracebacks which are much more readable
- `integration install` commands have a much nicer output as it installs integrations

There still are some parts that remain possible upgrades to this PR:

- conceal all the logging output for the `stack up` and `stack down` commands. Not sure to what extent we want to do that.
- logging -- it's pretty easy to switch out the default handler for a `RichHandler` but I'm not sure it's worth it right now.
- colour tweaks -- not sure how much we want to make those kinds of changes.
- using `rich.tree` as a way of visualising pipelines as DAGs (possibly...)

To get a feel for some of the new changes, try out these commands:

```shell
zenml init
zenml example list
zenml example info airflow_local
zenml example pull airflow_local
zenml integration list
zenml integration install
zenml stack list
zenml stack describe
zenml orchestrator list
zenml orchestrator describe
# and the list and describe commands apply to container_registry, artifact_store and metadata_store as well
```

AlexejPenner commented:
Currently trying it out locally - for me the formatting is slightly boken for `zenml integration list`
![brokenRich](https://user-images.githubusercontent.com/38859294/152569359-0ebfad94-1dca-456a-8b87-2eea6fc06a8b.png)

alex-zenml commented:
> Currently trying it out locally - for me the formatting is slightly boken for `zenml integration list` ![brokenRich](https://user-images.githubusercontent.com/38859294/152569359-0ebfad94-1dca-456a-8b87-2eea6fc06a8b.png)

where are you trying this?

AlexejPenner commented:
I tried running `zenml example info airflow_local`
![ZenmlExampleInfo](https://user-images.githubusercontent.com/38859294/152570553-bef3a795-b83d-4a4b-9457-e9d0634aa43d.png)


Also run in Pycharm Terminal

AlexejPenner commented on src/zenml/cli/integration.py:
```diff
@@ -42,29 +44,10 @@ def integration() -> None:
+    formatted_table = format_integration_list(
+        list(integration_registry.integrations.items())
     )
-    warning("zenml integration install EXAMPLE_NAME")
```
You think we no longer need to tell the user what he could do next?

alex-zenml replied:
Good point. Not sure why I thought that was worth removing. Have added it back in now.

AlexejPenner commented on src/zenml/logger.py:
```diff
@@ -119,6 +118,11 @@ def get_console_handler() -> Any:
     console_handler = logging.StreamHandler(sys.stdout)
     console_handler.setFormatter(CustomFormatter())
     return console_handler
+    # console_handler = RichHandler(
+    #     show_path=False, omit_repeated_times=False, console=console
+    # )
+    # console_handler.setFormatter(CustomFormatter())
+    # return console_handler
```
This seems to be forgotten code

alex-zenml replied:
Actually not. I wanted to leave it in in case we decide to tackle the logging part in the future. Maybe that's not such a great practice, though. Better to just delete?

AlexejPenner commented on src/zenml/stack/stack.py:
```diff
@@ -386,7 +386,7 @@ def resume(self) -> None:
                 resources.
         """
-        logger.info("Resuming provisioned resources for stack '%s'.", self.name)
+        logger.info(f"Resuming provisioned resources for stack '{self.name}'.")
```
I think where possible, we use %s in logger.

alex-zenml replied:
Reverted.

AlexejPenner commented on src/zenml/stack/stack.py:
```diff
@@ -403,7 +403,7 @@ def resume(self) -> None:
         """Suspends the provisioned local resources of the stack."""
         logger.info(
-            "Suspending provisioned resources for stack '%s'.", self.name
+            f"Suspending provisioned resources for stack '{self.name}'."
```
same as above

AlexejPenner reviewed (changes requested):
I like the changes where they work, unfortunately the emojis don't render on my machine

htahir1 commented on src/zenml/cli/utils.py:
```diff
@@ -117,19 +161,70 @@ def print_stack_component_list(
+    ]
+    console.print(rich_table)
+
+    # title(f"{display_name}:")
```
Remove these?

htahir1 commented on src/zenml/cli/utils.py:
```diff
@@ -206,11 +301,29 @@ def _wrapper(*args: Any, **kwargs: Any) -> Any:
     """Uninstalls pypi package from the current environment with pip"""
     subprocess.check_call(
-        [sys.executable, "-m", "pip", "uninstall", "-y", package]
+        [
```
Should this also have 

```
            "-qqq",
            "--no-warn-conflicts",
```
            
            like above?

htahir1 reviewed (changes requested):
Its a bit hard to review this more deeply for now. Left a few comments, can take another round at it later too. Really cool features!

stefannica reviewed (approved):
Great PR. I can't find anything to point out aside from what the others already mentioned. Nice work, can't wait to see this in action.

htahir1 commented on src/zenml/cli/example.py:
```diff
@@ -474,20 +477,20 @@ def clean(git_examples_handler: GitExamplesHandler, path: str) -> None:
-    #  pure .md dump)
     try:
         example_obj = git_examples_handler.get_examples(example_name)[0]
 
```
Not sure anymore, feel free to disregard

alex-zenml commented:
@stefannica @htahir1 @AlexejPenner I've addressed all comments / suggestions. (Thank you for those!) I don't know what to do about the weird formatting that Alexej brought up except for others in the team to checkout the branch and try to see if they have the same issues.

Once we've checked that, this PR should be good to go.

alex-zenml commented:
@htahir1 So to confirm, I'm good to merge it?
